### ISSUE TITLE
Evohome in Dashticz

### ISSUE THREAD

**Issue by lokonli** [Dashticz/dashticz#68]
https://www.domoticz.com/forum/viewtopic.php?f=67&t=29533

**Comment by wimmme** [comment]
I would absolutely love that too. Will see if I can code this myself but cannot make any promise.

**Comment by lokonli** [comment]
@wimmme : Working on this?

The json interface calls to Domoticz are documented here:
https://www.domoticz.com/wiki/Evohome#Notes

I don't have an Evohome so for me it's a bit difficult to develop/test.

I can help with the Dashticz structure. If you need help let me know.

**Comment by lokonli** [comment]
I've created a quick hack for testing evohome in a separate branch 'evohome':

https://github.com/Dashticz/dashticz/tree/evohome

Can someone test this?

The evohome zones (actual temperature + setpoint) should become visible, and it should be possible to temporary overrule the setpoint.

**Comment by clinkadink** [comment]
Hi lokonli, thanks for this. 

I tried it on and its doesn't load the main screen. It shows the "Loading Dashticz ...", then fades out ... but then nothing. Black screen.

Chrome dev tools shows this:

```
Uncaught TypeError: $.ajax(...).fail(...).then(...).then(...).then(...).then(...).then(...).then(...).then(...).then(...).then(...).catch is not a function
    at loadFiles (main.js?v=1577749344396:227)
```

I also noticed that the main.js seems quite a bit different from mine! 

<img width="481" alt="error" src="https://user-images.githubusercontent.com/33834551/71605166-e0672e00-2b5e-11ea-975e-c756902e7750.png">

**Comment by clinkadink** [comment]
Me again, I found 4 changes to your main.js, so I just tried copying them across, into my main.js. 

1. handleDevice at line 1219
2. getEvohomeBlock at 1874
3. addEvoThermostatFunctions at 1911
4. switchEvoZone at 1962

I saved and refreshed Dashticz and hey presto, the Evohome thermostat for my kitchen now has plus and minus controls! Its just gone midnight, so the setpoint is set to 15c, but the room is still toasty. 

![image](https://user-images.githubusercontent.com/33834551/71605490-ecec8600-2b60-11ea-9b8e-39037e0367ac.png)

Well done. Again, thanks for this. Great work.

The "Controls" on the left is the actual Evohome hub where I can set different modes. I will now try to get these available in Dashticz, like they do in Domoticz (shown below).

![image](https://user-images.githubusercontent.com/33834551/71605595-f0ccd800-2b61-11ea-878f-8499fed36656.png)

**Comment by lokonli** [comment]
The evohome work is derived from beta and should work. Did you try refreshing the cache?

Can you post the Domoticz JSON output of the control hub?

**Comment by clinkadink** [comment]
Yes, that and icognito. Still no luck. Failing on the `loadFiles` function.

Incidentally, I only cloned the master via git a couple of weeks ago, and started using Dashticz. But I notice your branch's main.js is quite different. I guess this is due to me using master, and not beta. This is my [main.js](https://github.com/Dashticz/dashticz/blob/master/js/main.js) and this is [yours](https://github.com/Dashticz/dashticz/blob/evohome/js/main.js).

**Comment by clinkadink** [comment]
I managed to merge it will my main.js, and got the rest working.

![image](https://user-images.githubusercontent.com/33834551/71644762-c1cf7700-2cc5-11ea-94a7-85e44818e6f9.png)

The Evohome controller is now allowing me to select the mode.

![image](https://user-images.githubusercontent.com/33834551/71644776-15da5b80-2cc6-11ea-804a-8e97d3ef52e6.png)

The `switchEvoZone ` url was missing the `until `parameter. So I added it, pulling in a default boost time, i.e. `settings['evohome_boost_zone']`, which adds x minutes to the current time.

I.e. from this ...
`/json.htm?username=' + usrEnc + '&password=' + pwdEnc + '&type=setused&idx=' + idx + '&setpoint=' + setpoint + '&mode=TemporaryOverride&used=true&jsoncallback=?'`

To this ...
`/json.htm?username=' + usrEnc + '&password=' + pwdEnc + '&type=setused&idx=' + idx + '&setpoint=' + setpoint + '&mode=TemporaryOverride&until='+moment().add(settings['evohome_boost_zone'], 'minutes').toISOString() +'&used=true&jsoncallback=?'`

The plus/minus buttons for each zone block obviously create a temporary override, diverting from the schedule. I added an icon to show whether the device is following the schedule, or has a temporary override.

![image](https://user-images.githubusercontent.com/33834551/71644895-dad92780-2cc7-11ea-9b97-c4d255adf862.png)

I also modified the function to allow cancelling the temporary override by clicking/touching the temporary override. This reverts the device to follow the schedule, i.e. `mode=Auto`

Finally, instead of displaying the time the device was last updated, `device['LastUpdate']`, I had added the next setpoint time, `device['Until']`. However, this will only work for `device['Status']` of either Auto or TemporaryOverride. Added some logic to catch this.

`var untilOrLastUpdate = (status == 'Auto' || status == 'TemporaryOverride') ? 'Until ' + moment(device['Until']).format('HH:mm') : moment(device['LastUpdate']).format(settings['timeformat']);`

Thanks again for this :)

**Comment by clinkadink** [comment]
My updates are [here](https://github.com/clinkadink/dashticz/blob/master/js/main.js).

**Comment by lokonli** [comment]
Nice work. I'll merge the changes later this week into beta.

**Comment by wimmme** [comment]
Darned, you guys are fast. I was on a new years holiday in 'offline' mode for a week, i come back and BANG, you got this !
Will test this as soon as I am home and the beta branch is updated. 
God I love this community.

**Comment by lokonli** [comment]
@clinkadink:
I've merged your changes into the evohome branch.

This branch is derived from beta, and should work. Can you test? 
After that I'll merge it into beta.

Still some work todo:
* Icon scaling is not correct (on my system)
* Some language files needs to be updated for the additional config settings.

I'll do that as a next step.

**Comment by clinkadink** [comment]
@lokonli 

Just cloned the evohome branch. Works fine. No issues. I can control my evohome devices perfectly. I have verified this against the Honeywell web app, and can see the mode, setpoints change seconds later. 

The Evohome controller and hot water png is currently missing, but you mentioned this.

Cheers again :)

**Comment by lokonli** [comment]
ok, great.
I'll add your images, update the documentation, and merge it into beta.

**Comment by lokonli** [comment]
Done.
Can you test latest beta? (3.3.0)

**Comment by clinkadink** [comment]
Just downloaded the beta (3.3.0), works a treat - no issues :)

![image](https://user-images.githubusercontent.com/33834551/71848300-093f7400-30c7-11ea-977b-bab97f43dbea.png)

The release notes look fine too.

https://dashticz.readthedocs.io/en/beta/blocks/domoticzblocks.html#evohome

I will now move across to this beta.

Many thanks for this, keep up the great work.

**Comment by wimmme** [comment]
Looking good for me too, nice work !

**Comment by lokonli** [comment]
Merged into beta.

**Comment by JanvdW304** [comment]
Nice work! Last weekend I updated dashticz and saw this great functionality. I was wondering why in my case the device['LastUpdate'] was still presented.

It appeared that this is the case when temperature is set permanent (I set temperatures with scripts, based on presence switches). Is it possible to show 'Permanent' (or nothing at all) in this case? Or make it dependent on the 'last_update' parameter for a device in CONFIG.js?
